crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs

```diff
@@ -0,0 +1,566 @@
+        final_block
+    }
+
+    /// Inline the ending block of a branch, the point where all blocks from a jmpif instruction
+    /// join back together. In particular this function must handle merging block arguments from
+    /// all of the join point's predecessors, and it must handle any differing side effects from
+    /// each branch.
+    ///
+    /// Afterwards, continues inlining recursively until it finds the next end block or finds the
+    /// end of the function.
+    ///
+    /// Returns the final block that was inlined.
+    fn inline_branch_end(
+        &mut self,
+        destination: BasicBlockId,
+        condition: ValueId,
+        last_then_block: BasicBlockId,
+        last_else_block: BasicBlockId,
+    ) -> BasicBlockId {
+        assert_eq!(self.cfg.predecessors(destination).len(), 2);
+
+        let then_args = self.function.dfg[last_then_block].terminator_arguments();
+        let else_args = self.function.dfg[last_else_block].terminator_arguments();
+
+        let params = self.function.dfg.block_parameters(destination);
+        assert_eq!(params.len(), then_args.len());
+        assert_eq!(params.len(), else_args.len());
+
+        let args = vecmap(then_args.iter().zip(else_args), |(then_arg, else_arg)| {
+            (self.translate_value(*then_arg), self.translate_value(*else_arg))
+        });
+
+        // Cannot include this in the previous vecmap since it requires exclusive access to self
+        let args =
+            vecmap(args, |(then_arg, else_arg)| self.merge_values(condition, then_arg, else_arg));
+
+        // insert merge instruction
+        self.inline_block(destination, &args)
+    }
+
+    /// Inline all instructions from the given destination block into the entry block.
+    /// Afterwards, check the block's terminator and continue inlining recursively.
+    ///
+    /// Returns the final block that was inlined.
+    ///
+    /// Expects that the `arguments` given are already translated via self.translate_value.
+    /// If they are not, it is possible some values which no longer exist, such as block
+    /// parameters, will be kept in the program.
+    fn inline_block(&mut self, destination: BasicBlockId, arguments: &[ValueId]) -> BasicBlockId {
+        let parameters = self.function.dfg.block_parameters(destination);
+        Self::insert_new_instruction_results(
+            &mut self.values,
+            parameters,
+            InsertInstructionResult::Results(arguments),
+        );
+
+        // If this is not a separate variable, clippy gets confused and says the to_vec is
+        // unnecessary, when removing it actually causes an aliasing/mutability error.
+        let instructions = self.function.dfg[destination].instructions().to_vec();
+        for instruction in instructions {
+            self.push_instruction(instruction);
+        }
+
+        self.handle_terminator(destination)
+    }
+
+    fn push_instruction(&mut self, id: InstructionId) {
```

Comment:
Doc comment please - particularly that the insert is always to the entry block

Reply:
Added